File: C++/크레인인형뽑기게임.cpp
#include <string>
#include <vector>

using namespace std;

int solution(vector<vector<int>> board, vector<int> moves) {
    int N = board.size();
    vector<int> top_idx(N);
    for(int i = 0; i < N; i++){
        for(int j = 0; j < N; j++){
            if(board[j][i] != 0){
                top_idx[i] = j;
                break;
            }
        }
    }
    
    int answer = 0;
    vector<int> basket;
    for(int m = 0; m < moves.size(); m++){
        int move = moves[m]-1;
        
         if(top_idx[move] == N){
            continue;
        }
        if(!basket.empty() && basket.back() == board[top_idx[move]][move]){
            basket.pop_back();
            answer += 2;
        }
        else{
            basket.push_back(board[top_idx[move]][move]);
        }
        top_idx[move]++;
    }
    
    return answer;
}

File: level1/크레인인형뽑기게임.py
def solution(board, moves):
    top_idx = []
    N = len(board)
    for j in range(N):
        flag = False
        for i in range(N):
            if board[i][j] != 0:
                flag = True
                top_idx.append(i)
                break
        if not flag:
            top_idx.append(N)
    
    answer = 0
    stack = []
    for m in moves:
        if top_idx[m-1] >= N:
            continue
            
        if stack and stack[-1] == board[top_idx[m-1]][m-1]:
            answer += 2
            stack.pop()
        else:
            stack.append(board[top_idx[m-1]][m-1])
            
        board[top_idx[m-1]][m-1] = 0
        top_idx[m-1] += 1
    
    return answer
